provider/providertest/integration.go

```diff
@@ -0,0 +1,422 @@
+	}
+
+	err = fetch(providerCreator, &resource)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	log.Printf("%s verify fields\n", resource.Table.Name)
+	conn, err := setupDatabase()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer conn.Close(context.Background())
+
+	err = verifyFields(resource, conn)
+	if err != nil {
+		t.Fatal(err)
+	}
+}
+
+// setup - puts *.tf files into isolated test dir and creates the instance of terraform
+func setup(resource *ResourceIntegrationTestData) (*tfexec.Terraform, error) {
+	w, err := copyTfFiles(resource.Table.Name)
+	if err != nil {
+		// remove workdir
+		if e := os.RemoveAll(resource.workdir); e != nil {
+			return nil, fmt.Errorf("failed to RemoveAll after: %s\n reason:%s", err, e)
+		}
+		return nil, err
+	}
+	resource.workdir = w
+
+	lookPath := os.Getenv("TF_EXEC_PATH")
+	if lookPath == "" {
+		lookPath = "terraform"
+	}
+	execPath, err := exec.LookPath(lookPath)
+	if err != nil {
+		return nil, err
+	}
+	tf, err := tfexec.NewTerraform(resource.workdir, execPath)
+	if err != nil {
+		return nil, err
+	}
+	if err = enableTerraformLog(tf, resource.workdir); err != nil {
+		return nil, err
+	}
+	return tf, nil
+}
+
+// deploy - uses terraform to deploy resources and builds teardown function. deployment timeout can be set via TF_EXEC_TIMEOUT env variable
+func deploy(tf *tfexec.Terraform, resource *ResourceIntegrationTestData) (func() error, error) {
+	// prepare terraform variables
+	hostname, err := os.Hostname()
+	if err != nil {
+		return nil, err
+	}
+
+	resource.Suffix = simplifyString(hostname)
+	resource.Prefix = simplifyString(resource.Table.Name)
+	testSuffix := fmt.Sprintf("test_suffix=%s", resource.Suffix)
+	testPrefix := fmt.Sprintf("test_prefix=%s", resource.Prefix)
+
+	teardown := func() error {
+		err = tf.Destroy(context.Background(), tfexec.Var(testPrefix),
+			tfexec.Var(testSuffix))
+		if err != nil {
+			return err
+		}
+		log.Printf("%s cleanup\n", resource.Table.Name)
+		if err := os.RemoveAll(resource.workdir); err != nil {
+			return err
+		}
+		log.Printf("%s done\n", resource.Table.Name)
+		return nil
+	}
+
+	ctx := context.Background()
+	if i, err := strconv.Atoi(os.Getenv("TF_EXEC_TIMEOUT")); err == nil {
+		var cancel context.CancelFunc
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(i)*time.Minute)
+		defer cancel()
+	}
+
+	log.Printf("%s tf init\n", resource.Table.Name)
+	if err = tf.Init(ctx, tfexec.Upgrade(true)); err != nil {
+		return teardown, err
+	}
+
+	ticker := time.NewTicker(5 * time.Minute)
+	startTime := time.Now()
+	applyDone := make(chan bool)
+	go func() {
+		for {
+			select {
+			case <-applyDone:
+				return
+			case timestamp := <-ticker.C:
+				log.Printf("%s applying for %v", resource.Table.Name, timestamp.Sub(startTime))
+			}
+		}
+	}()
+
+	log.Printf("%s tf apply -var test_prefix=%s -var test_suffix=%s\n", resource.Table.Name, resource.Prefix, resource.Suffix)
+	if err = tf.Apply(ctx, tfexec.Var(testPrefix), tfexec.Var(testSuffix)); err != nil {
```

Comment:
if you get a timeout here the goroutine won't stop, you need to add in the select <-ctx.Done()